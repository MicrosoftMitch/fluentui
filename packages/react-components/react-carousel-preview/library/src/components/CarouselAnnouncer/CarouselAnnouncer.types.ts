import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import * as React from 'react';

export type CarouselAnnouncerSlots = {
  root: Slot<'div'>;
};

export type AnnouncerIndexRenderFunction = (
  index: number,
  totalSlides: number,
  slideGroupList: number[][],
) => React.ReactNode;
/**
 * CarouselAnnouncer Props
 */
export type CarouselAnnouncerProps = Omit<ComponentProps<Partial<CarouselAnnouncerSlots>>, 'children'> & {
  children: AnnouncerIndexRenderFunction;
};

/**
 * State used in rendering CarouselAnnouncer
 */
export type CarouselAnnouncerState = ComponentState<CarouselAnnouncerSlots> & {
  /**
   * The function that will render nav items based on total slides and their index.
   */
  renderAnnouncerChild: AnnouncerIndexRenderFunction;

  /**
   * The total number of slides passed in from carousel context.
   */
  totalSlides: number;

  /**
   * The current index passed in from carousel context.
   */
  currentIndex: number;

  /**
   * The list of cards in each slide based on index.
   */
  slideGroupList: number[][];
};
