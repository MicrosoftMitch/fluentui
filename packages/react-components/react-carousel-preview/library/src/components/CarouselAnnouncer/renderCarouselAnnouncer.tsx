/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { CarouselAnnouncerState, CarouselAnnouncerSlots } from './CarouselAnnouncer.types';

/**
 * Render the final JSX of CarouselAnnouncer
 */
export const renderCarouselAnnouncer_unstable = (state: CarouselAnnouncerState) => {
  assertSlots<CarouselAnnouncerSlots>(state);

  return <state.root />;
};
