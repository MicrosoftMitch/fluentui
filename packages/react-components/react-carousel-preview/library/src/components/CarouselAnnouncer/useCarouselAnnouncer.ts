import * as React from 'react';
import { getIntrinsicElementProps, slot, useIsomorphicLayoutEffect, useMergedRefs } from '@fluentui/react-utilities';
import type { CarouselAnnouncerProps, CarouselAnnouncerState } from './CarouselAnnouncer.types';
import { useCarouselContext_unstable as useCarouselContext } from '../CarouselContext';

/**
 * Create the state required to render CarouselAnnouncer.
 *
 * The returned state can be modified with hooks such as useCarouselAnnouncerStyles_unstable,
 * before being passed to renderCarouselAnnouncer_unstable.
 *
 * @param props - props from this instance of CarouselAnnouncer
 * @param ref - reference to root HTMLDivElement of CarouselAnnouncer
 */
export const useCarouselAnnouncer_unstable = (
  props: CarouselAnnouncerProps,
  ref: React.Ref<HTMLDivElement>,
): CarouselAnnouncerState => {
  const subscribeForValues = useCarouselContext(ctx => ctx.subscribeForValues);
  const textRef: React.Ref<HTMLElement> = React.useRef(null);

  const currentIndex = useCarouselContext(ctx => ctx.activeIndex);
  const navLengthRef = React.useRef<number>(0);
  const navGroupRef = React.useRef<number[][]>([]);

  const updateInnerText = (index: number, totalSlides: number, groupIndexList: number[][]) => {
    if (totalSlides <= 0) {
      // Ignore announcements until slides discovered
      return;
    }
    const announcementText = props.children(index, totalSlides, groupIndexList);

    if (textRef.current) {
      textRef.current.innerText = announcementText;
    }
  };

  useIsomorphicLayoutEffect(() => {
    return subscribeForValues(data => {
      navLengthRef.current = data.navItemsCount;
      navGroupRef.current = data.groupIndexList;
      updateInnerText(data.activeIndex, data.navItemsCount, data.groupIndexList);
    });
  }, [subscribeForValues, updateInnerText]);

  useIsomorphicLayoutEffect(() => {
    updateInnerText(currentIndex, navLengthRef.current, navGroupRef.current);
  }, [currentIndex, updateInnerText]);

  return {
    renderAnnouncerText: props.children,
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref: useMergedRefs(ref, textRef),
        ...props,
        children: null,
      }),
      {
        elementType: 'div',
        defaultProps: {
          'aria-live': 'polite',
        },
      },
    ),
  };
};
