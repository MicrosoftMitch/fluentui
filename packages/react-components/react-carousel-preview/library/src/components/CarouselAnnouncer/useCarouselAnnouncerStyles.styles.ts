import { makeStyles, mergeClasses } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type { CarouselAnnouncerSlots, CarouselAnnouncerState } from './CarouselAnnouncer.types';

export const carouselAnnouncerClassNames: SlotClassNames<CarouselAnnouncerSlots> = {
  root: 'fui-CarouselAnnouncer',
  // TODO: add class names for all slots on CarouselAnnouncerSlots.
  // Should be of the form `<slotName>: 'fui-CarouselAnnouncer__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    // TODO Add default styles for the root element
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the CarouselAnnouncer slots based on the state
 */
export const useCarouselAnnouncerStyles_unstable = (state: CarouselAnnouncerState): CarouselAnnouncerState => {
  'use no memo';

  const styles = useStyles();
  state.root.className = mergeClasses(carouselAnnouncerClassNames.root, styles.root, state.root.className);

  // TODO Add class names to slots, for example:
  // state.mySlot.className = mergeClasses(styles.mySlot, state.mySlot.className);

  return state;
};
