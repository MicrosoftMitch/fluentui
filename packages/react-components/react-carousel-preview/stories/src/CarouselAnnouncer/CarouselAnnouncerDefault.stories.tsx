import * as React from 'react';
import { CarouselAnnouncer, CarouselAnnouncerProps } from '@fluentui/react-carousel-preview';

export const Default = (props: Partial<CarouselAnnouncerProps>) => <CarouselAnnouncer {...props} />;
