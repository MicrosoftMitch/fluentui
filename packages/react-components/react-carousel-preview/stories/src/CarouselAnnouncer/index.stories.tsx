import { CarouselAnnouncer } from '@fluentui/react-carousel-preview';

import descriptionMd from './CarouselAnnouncerDescription.md';
import bestPracticesMd from './CarouselAnnouncerBestPractices.md';

export { Default } from './CarouselAnnouncerDefault.stories';

export default {
  title: 'Preview Components/CarouselAnnouncer',
  component: CarouselAnnouncer,
  parameters: {
    docs: {
      description: {
        component: [descriptionMd, bestPracticesMd].join('\n'),
      },
    },
  },
};
