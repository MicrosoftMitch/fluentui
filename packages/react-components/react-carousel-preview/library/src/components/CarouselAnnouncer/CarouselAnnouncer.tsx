import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCarouselAnnouncer_unstable } from './useCarouselAnnouncer';
import { renderCarouselAnnouncer_unstable } from './renderCarouselAnnouncer';
import { useCarouselAnnouncerStyles_unstable } from './useCarouselAnnouncerStyles.styles';
import type { CarouselAnnouncerProps } from './CarouselAnnouncer.types';

/**
 * CarouselAnnouncer component - TODO: add more docs
 */
export const CarouselAnnouncer: ForwardRefComponent<CarouselAnnouncerProps> = React.forwardRef((props, ref) => {
  const state = useCarouselAnnouncer_unstable(props, ref);

  useCarouselAnnouncerStyles_unstable(state);

  /**
   * @see https://github.com/microsoft/fluentui/blob/master/docs/react-v9/contributing/rfcs/react-components/convergence/custom-styling.md
   *
   * TODO: ðŸ’¡ once package will become stable (PR which will be part of promoting PREVIEW package to STABLE),
   *      - uncomment this line
   *      - update types {@link file://./../../../../../../../packages/react-components/react-shared-contexts/library/src/CustomStyleHooksContext/CustomStyleHooksContext.ts#CustomStyleHooksContextValue}
   *      - verify that custom global style override works for your component
   */
  // useCustomStyleHook_unstable('useCarouselAnnouncerStyles_unstable')(state);

  return renderCarouselAnnouncer_unstable(state);
});

CarouselAnnouncer.displayName = 'CarouselAnnouncer';
