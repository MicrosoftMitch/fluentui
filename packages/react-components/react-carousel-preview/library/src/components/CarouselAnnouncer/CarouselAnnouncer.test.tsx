import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { CarouselAnnouncer } from './CarouselAnnouncer';

describe('CarouselAnnouncer', () => {
  isConformant({
    Component: CarouselAnnouncer,
    displayName: 'CarouselAnnouncer',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<CarouselAnnouncer>{() => 'Default CarouselAnnouncer'}</CarouselAnnouncer>);
    expect(result.container).toMatchSnapshot();
  });
});
