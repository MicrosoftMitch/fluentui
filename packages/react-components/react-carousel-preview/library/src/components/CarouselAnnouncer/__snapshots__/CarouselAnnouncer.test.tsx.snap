// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselAnnouncer renders a default state 1`] = `
<div>
  <div
    class="fui-CarouselAnnouncer"
  >
    Default CarouselAnnouncer
  </div>
</div>
`;
